crates/pera-core/src/signature_mpc/mpc_manager.rs

```diff
@@ -2,163 +2,231 @@ use crate::authority::authority_per_epoch_store::AuthorityPerEpochStore;
-        self.spawn_mpc_messages_handler(public_parameters, messages_handler_receiver);
+    async fn new(
+        consensus_adapter: Arc<dyn SubmitToConsensus>,
+        epoch_store: Weak<AuthorityPerEpochStore>,
+        mpc_threshold_number_of_parties: usize,
+        session_id: ObjectID,
+        public_parameters: ProofPublicParameters,
+    ) -> Self {
+        let mut new_instance = Self {
+            status: MPCSessionStatus::Active,
+            pending_messages: HashMap::new(),
+            consensus_adapter: consensus_adapter.clone(),
+            epoch_store: epoch_store.clone(),
+            mpc_threshold_number_of_parties,
+            session_id,
+            party: None,
+        };
+        match new_instance.start_proof_mpc_flow(public_parameters).await {
+            Ok(party) => {
+                new_instance.party = Some(party);
+                new_instance
+            }
+            Err(err) => {
+                // This should never happen, as there should be on-chain verification on the init transaction, and
+                // we are ignoring failed transactions.
+                error!("Error initializing the MPC proof flow: {:?}", err);
+                new_instance.status = MPCSessionStatus::Finished;
+                new_instance
+            }
+        }
     }
 
-    /// Spawns an asynchronous task to handle incoming messages.
-    /// The [`MPCService`] will forward any message related to that instance to this channel.
-    fn spawn_mpc_messages_handler(
-        &self,
-        public_parameters: ProofPublicParameters,
-        mut receiver: mpsc::Receiver<MPCInput>,
-    ) {
-        let consensus_adapter = Arc::clone(&self.consensus_adapter);
-        let epoch_store = self.epoch_store.clone();
-        let threshold = self.mpc_threshold_number_of_parties;
-        tokio::spawn(async move {
-            let mut party: ProofParty;
-            while let Some(message) = receiver.recv().await {
-                match message {
-                    MPCInput::InitEvent(_) => {
-                        party = match Self::handle_mpc_proof_init_event(
-                            public_parameters.clone(),
-                            consensus_adapter.clone(),
-                            epoch_store.clone(),
-                            threshold,
-                        )
-                        .await
-                        {
-                            Ok(party) => party,
-                            Err(err) => {
-                                // This should never happen, as there should be on-chain verification on the init transaction
-                                return;
-                            }
-                        };
-                    }
-                    MPCInput::Message => {
-                        // TODO (#235): Implement MPC messages handling
-                    }
-                }
+    fn advance(&mut self) -> Option<ConsensusTransaction> {
```

Comment:
Can we have some documentation describing this?

Reply:
Also, it's weird that "advancing an MPC Instance" returns something of the type `ConsensusTransaction`; can we have
maybe a type alias here?